data-analytics/iot-nirvana/client/src/main/java/com/google/cloud/demo/iot/nirvana/client/MessagePublisher.java

```diff
@@ -0,0 +1,194 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.PKCS8EncodedKeySpec;
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
+import org.joda.time.DateTime;
+
+/** A MQTT client sending temperature data to a device registry on Google Cloud IoT Core */
+@AutoValue
+public abstract class MessagePublisher {
+
+  static final Logging logging = LoggingOptions.getDefaultInstance().getService();
+
+  private static double MIN_DELTA = -0.05;
+  private static double MAX_DELTA = 0.05;
+  private static final String CLOUD_IOT_CORE_MQTT_BRIDGE_HOST_NAME = "mqtt.googleapis.com";
+  private static final short CLOUD_IOT_CORE_MQTT_BRIDGE_PORT = 8883;
+
+  public abstract String getDeviceId();
+
+  public abstract ClientOptions getOptions();
+
+  public abstract CityTemperature getCityTemperature();
+
+  /** Builder for MessageSender. */
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setDeviceId(String deviceId);
+
+    public abstract Builder setOptions(ClientOptions options);
+
+    public abstract Builder setCityTemperature(CityTemperature cityTemperature);
+
+    public abstract MessagePublisher build();
+  }
+
+  public static Builder newBuilder() {
+    return new AutoValue_MessagePublisher.Builder();
+  }
+
+  /**
+   * Generate and publish a temperature messages to Cloud IoT Core with a 10% variance around a
+   * given temperature value and with a frequency of 1 message/second.
+   */
+  public void publishMessages() throws ClientException, TemperatureException {
+    try {
+      String mqttTopic = generateMqttTopic();
+      MqttClient client = createAndConnectMqttClient();
+      while (true) {
+        // Create the temperature message to publish
+        double variance = MIN_DELTA + (Math.random() * MAX_DELTA);
+        Message msg = new Message();
+        msg.setId(getCityTemperature().getLat(), getCityTemperature().getLng());
+        msg.setTemperature(getCityTemperature().getTemperature() + variance);
+        msg.setTimestamp(new java.util.Date().getTime());
+
+        // Create the MQTT message (in this case QoS=1 --> delivered at least once)
+        String payload = new Gson().toJson(msg);
+        MqttMessage message = new MqttMessage(payload.getBytes());
+        message.setQos(1);
+
+        // Publish the message and sleep for 1 second
+        client.publish(mqttTopic, message);
+        logDebug("Message published: " + message);
+        try {
+          Thread.sleep(1000);
+        } catch (InterruptedException e) {
+          logDebug("Message sending interrupted.");
```

Comment:
I am not sure if this will be useful for debugging? The intent here is to just loop with 1 second delays..

Reply:
Removed in commit aa68e7f20fc4feb6103d063226e7caeb3f4cc266